primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,468 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Offchain Storage Lock
+//!
+//! A storage-based lock with a defined expiry time.
+//!
+//! The lock is using Local Storage and allows synchronizing access to critical
+//! section of your code for concurrently running Off-chain Workers. Usage of
+//! `PERSISTENT` variant of the storage persists the lock value accross a full node
+//! restart or re-orgs.
+//!
+//! A use case for the lock is to make sure that a particular section of the
+//! code is only run by one Off-chain Worker at the time. This may include
+//! performing a side-effect (i.e. an HTTP call) or alteration of single or
+//! multiple Local Storage entries.
+//!
+//! One use case would be collective updates of multiple data items or append /
+//! remove of i.e. sets, vectors which are stored in the offchain storage DB.
+//!
+//! ## Example:
+//!
+//! ```rust
+//! # use crate::offchain::storage::StorageValueRef;
+//! # use codec::{Decode, Encode, Codec};
+//! // in your off-chain worker code
+//!
+//! fn append_to_in_storage_vec<'a, T>(key: &'a [u8], _: T) where T: Encode {
+//!    // `access::lock` defines the storage entry which is used for
+//!    // persisting the lock in the underlying database.
+//!    // The entry name _must_ be unique and can be seen as mutex instance reference.
+//!    let mut lock = StorageLock::new(b"access::lock");
+//!    {
+//!         let _guard = lock.spin_lock();
+//!         let acc = StorageValueRef::persistent(key);
+//!         let v: Vec<T> = acc.get::<Vec<T>>().unwrap().unwrap();
+//!         // modify `v` as desired
+//!         // i.e. perform some heavy computation
+//!         // side effects that should only be done once.
+//!         acc.set(v);
+//!         // drop `_guard` implicitly at end of scope
+//!    }
+//! }
+//! ```
+
+use crate::offchain::storage::StorageValueRef;
+use codec::{Codec, Decode, Encode};
+use sp_core::offchain::{Duration, Timestamp};
+use sp_io::offchain;
+
+//use frame_system how?
+
+/// Default expiry duration in milliseconds.
+const STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS: u64 = 30_000;
+
+/// Snooze duration before attempting to lock again in ms.
+const STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE: u64 = 100;
+
+/// Lockable item for use with a persisted storage lock.
+///
+/// Bound for an item that has a stateful ordered meaning
+/// without explicitly requiring `Ord` trait in general.
+pub trait Lockable: Sized + Codec + Copy {
```

Comment:
Not a fan of static vs instance methods here, why not:
```
pub trait Lockable: Sized + Codec + Copy + PartialOrd {
  fn current() -> Self;
  fn deadline() -> Self;
  fn snooze();
}
```
Seems that `snooze` is being called on `L::current()` anyway.

Reply:
`fn snooze()` takes the existing in-db deadline value, thus it's an argument.

Reply:
Well, that's my point that with couple of small changes that I suggested you can keep the `Lockable` trait methods fully static, and reduce the number of methods to 3 by introducing `PartialOrd`.

Having `snooze` static makes the default implementation much more reasonable (since it's static anyway) and the only use case in our code to use `snooze` is always: `L::current().snooze()` so instead we can change the static implementation for both `BlockAndTime` and `Time` locks to call `current` internally since they override the default.

The `expired(&self)/snooze(&self)` methods is not super readable imho - it creates extra cognitive load to figure out what two things do we compare and in what direction. I.e. which one is correct:
```
deadline.expired(now)
// or 
now.expired(deadline)
```
or: 
```
deadline.snooze(now)
// vs
now.snooze(deadline)
```
How long will I snooze in both cases?
Compared with:
```
L::snooze(deadline)
// or even
L::snooze_until(deadline)
```
This imho makes it much easier to understand especially given that `Lockable` has a notion of some absolute, monotonic value (like time or block numbers) - so it's clear that single instance should be enough to indicate what you want to do.
Now I feel like `Lockable` is a mix of `Instant` and `Duration` - the API requires you to provide two `Instants` for some reason to be able to snooze which imho causes confusion.

Reply:
I think I missunderstod you here: I thought you wanted to get rid of the deadline argument.. which made no sense to me (: - with the more detailed explanation I am with you.

Reply:
Note that I wanted to avoid `PartialOrd` for a couple of reasons: How would one handle a return value of `None`? I.e. for `lock(..)`, should that cause an expiration or continue the loop?

If we would decide for `Ord` to avoid the above, there would remain another issue:
A common assumption for implementations of `Ord` is commutativity on the users end.
Now combine this with compound `Instant`s (i.e. `Timestamp` and `BlockNumber`) then one ends up with smth like:
```
impl<B: BlockNumberProvider> core::cmp::PartialOrd for BlockAndTimeInstant<B> {
	use core::cmp::Ordering;
	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
		// attention! not cummutative, `x < y` does not imply `x > y`!
		if self.block_number > other.block_number && self.timestamp > other.timestamp {
			Ordering::Greater
		} else if self.block_number == other.block_number && self.timestamp == other.timestamp {
			Ordering::Equal
		} else {
			Ordering::Less
		}
	}
}
```
hence the expired function which is explicit about having no commutativity guarantee but _must_ provide a `bool` indication.

Reply:
Given the off-gh discussion the fn will be `fn has_expired(deadline: _) { .. }`